### ISSUE TITLE
Whitelist seems to ignore root namespace classes sometimes

### ISSUE THREAD

**Issue by sebastianfeldmann** [humbug/php-scoper#261]
## Bug report

| Question    | Answer
| ------------| ---------------
| Box version | PHP Scoper version 0.9.2
| PHP version | PHP 7.2.9
| Platform with version | MacOS
| Github Repo | - https://github.com/sebastianbergmann/phpunit/pull/3086/files

While working on the pull request above I noticed that `php-scoper` seems to have an issue with whitelisting classes in the root namespace.

In this example the `PHP_Token` classes get whitelisted correctly but the `PHPUnit_Framework_MockObject_MockObject` class gets prefixed although it is whitelisted. 

<details>
 <summary>scoper.inc.php</summary>
 
 ```php
<?php
declare(strict_types=1);
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

$whitelistClasses = [
    'PHPUnit\*',
    'PHPUnit_Framework_MockObject_MockObject',
    'SebastianBergmann\CodeCoverage\*',
    'PharIo\*',
    'PHP_Token',
    'PHP_TokenWithScope',
    'PHP_TokenWithScopeAndVisibility',
    'PHP_Token_ABSTRACT',
    'PHP_Token_AMPERSAND',
    'PHP_Token_AND_EQUAL',
    'PHP_Token_ARRAY',
    'PHP_Token_ARRAY_CAST',
    'PHP_Token_AS',
    'PHP_Token_AT',
    'PHP_Token_BACKTICK',
    'PHP_Token_BAD_CHARACTER',
    'PHP_Token_BOOLEAN_AND',
    'PHP_Token_BOOLEAN_OR',
    'PHP_Token_BOOL_CAST',
    'PHP_Token_BREAK',
    'PHP_Token_CALLABLE',
    'PHP_Token_CARET',
    'PHP_Token_CASE',
    'PHP_Token_CATCH',
    'PHP_Token_CHARACTER',
    'PHP_Token_CLASS',
    'PHP_Token_CLASS_C',
    'PHP_Token_CLASS_NAME_CONSTANT',
    'PHP_Token_CLONE',
    'PHP_Token_CLOSE_BRACKET',
    'PHP_Token_CLOSE_CURLY',
    'PHP_Token_CLOSE_SQUARE',
    'PHP_Token_CLOSE_TAG',
    'PHP_Token_COALESCE',
    'PHP_Token_COLON',
    'PHP_Token_COMMA',
    'PHP_Token_COMMENT',
    'PHP_Token_CONCAT_EQUAL',
    'PHP_Token_CONST',
    'PHP_Token_CONSTANT_ENCAPSED_STRING',
    'PHP_Token_CONTINUE',
    'PHP_Token_CURLY_OPEN',
    'PHP_Token_DEC',
    'PHP_Token_DECLARE',
    'PHP_Token_DEFAULT',
    'PHP_Token_DIR',
    'PHP_Token_DIV',
    'PHP_Token_DIV_EQUAL',
    'PHP_Token_DNUMBER',
    'PHP_Token_DO',
    'PHP_Token_DOC_COMMENT',
    'PHP_Token_DOLLAR',
    'PHP_Token_DOLLAR_OPEN_CURLY_BRACES',
    'PHP_Token_DOT',
    'PHP_Token_DOUBLE_ARROW',
    'PHP_Token_DOUBLE_CAST',
    'PHP_Token_DOUBLE_COLON',
    'PHP_Token_DOUBLE_QUOTES',
    'PHP_Token_ECHO',
    'PHP_Token_ELLIPSIS',
    'PHP_Token_ELSE',
    'PHP_Token_ELSEIF',
    'PHP_Token_EMPTY',
    'PHP_Token_ENCAPSED_AND_WHITESPACE',
    'PHP_Token_ENDDECLARE',
    'PHP_Token_ENDFOR',
    'PHP_Token_ENDFOREACH',
    'PHP_Token_ENDIF',
    'PHP_Token_ENDSWITCH',
    'PHP_Token_ENDWHILE',
    'PHP_Token_END_HEREDOC',
    'PHP_Token_EQUAL',
    'PHP_Token_EVAL',
    'PHP_Token_EXCLAMATION_MARK',
    'PHP_Token_EXIT',
    'PHP_Token_EXTENDS',
    'PHP_Token_FILE',
    'PHP_Token_FINAL',
    'PHP_Token_FINALLY',
    'PHP_Token_FOR',
    'PHP_Token_FOREACH',
    'PHP_Token_FUNCTION',
    'PHP_Token_FUNC_C',
    'PHP_Token_GLOBAL',
    'PHP_Token_GOTO',
    'PHP_Token_GT',
    'PHP_Token_HALT_COMPILER',
    'PHP_Token_IF',
    'PHP_Token_IMPLEMENTS',
    'PHP_Token_INC',
    'PHP_Token_INCLUDE',
    'PHP_Token_INCLUDE_ONCE',
    'PHP_Token_INLINE_HTML',
    'PHP_Token_INSTANCEOF',
    'PHP_Token_INSTEADOF',
    'PHP_Token_INTERFACE',
    'PHP_Token_INT_CAST',
    'PHP_Token_ISSET',
    'PHP_Token_IS_EQUAL',
    'PHP_Token_IS_GREATER_OR_EQUAL',
    'PHP_Token_IS_IDENTICAL',
    'PHP_Token_IS_NOT_EQUAL',
    'PHP_Token_IS_NOT_IDENTICAL',
    'PHP_Token_IS_SMALLER_OR_EQUAL',
    'PHP_Token_Includes',
    'PHP_Token_LINE',
    'PHP_Token_LIST',
    'PHP_Token_LNUMBER',
    'PHP_Token_LOGICAL_AND',
    'PHP_Token_LOGICAL_OR',
    'PHP_Token_LOGICAL_XOR',
    'PHP_Token_LT',
    'PHP_Token_METHOD_C',
    'PHP_Token_MINUS',
    'PHP_Token_MINUS_EQUAL',
    'PHP_Token_MOD_EQUAL',
    'PHP_Token_MULT',
    'PHP_Token_MUL_EQUAL',
    'PHP_Token_NAMESPACE',
    'PHP_Token_NEW',
    'PHP_Token_NS_C',
    'PHP_Token_NS_SEPARATOR',
    'PHP_Token_NUM_STRING',
    'PHP_Token_OBJECT_CAST',
    'PHP_Token_OBJECT_OPERATOR',
    'PHP_Token_OPEN_BRACKET',
    'PHP_Token_OPEN_CURLY',
    'PHP_Token_OPEN_SQUARE',
    'PHP_Token_OPEN_TAG',
    'PHP_Token_OPEN_TAG_WITH_ECHO',
    'PHP_Token_OR_EQUAL',
    'PHP_Token_PAAMAYIM_NEKUDOTAYIM',
    'PHP_Token_PERCENT',
    'PHP_Token_PIPE',
    'PHP_Token_PLUS',
    'PHP_Token_PLUS_EQUAL',
    'PHP_Token_POW',
    'PHP_Token_POW_EQUAL',
    'PHP_Token_PRINT',
    'PHP_Token_PRIVATE',
    'PHP_Token_PROTECTED',
    'PHP_Token_PUBLIC',
    'PHP_Token_QUESTION_MARK',
    'PHP_Token_REQUIRE',
    'PHP_Token_REQUIRE_ONCE',
    'PHP_Token_RETURN',
    'PHP_Token_SEMICOLON',
    'PHP_Token_SL',
    'PHP_Token_SL_EQUAL',
    'PHP_Token_SPACESHIP',
    'PHP_Token_SR',
    'PHP_Token_SR_EQUAL',
    'PHP_Token_START_HEREDOC',
    'PHP_Token_STATIC',
    'PHP_Token_STRING',
    'PHP_Token_STRING_CAST',
    'PHP_Token_STRING_VARNAME',
    'PHP_Token_SWITCH',
    'PHP_Token_Stream',
    'PHP_Token_Stream_CachingFactory',
    'PHP_Token_THROW',
    'PHP_Token_TILDE',
    'PHP_Token_TRAIT',
    'PHP_Token_TRAIT_C',
    'PHP_Token_TRY',
    'PHP_Token_UNSET',
    'PHP_Token_UNSET_CAST',
    'PHP_Token_USE',
    'PHP_Token_USE_FUNCTION',
    'PHP_Token_VAR',
    'PHP_Token_VARIABLE',
    'PHP_Token_WHILE',
    'PHP_Token_WHITESPACE',
    'PHP_Token_XOR_EQUAL',
    'PHP_Token_YIELD',
    'PHP_Token_YIELD_FROM',
];

return [
    'whitelist' => $whitelistClasses,
];
 ```
</details>

<details>
 <summary>Output</summary>
 
 ```bash
 $ php-scoper add-prefix --no-ansi --config build/scoper.inc.php --no-interaction --stop-on-failure --output-dir build/phar-prefixed --prefix PHPUnit build/phar
      ____  __  ______     _____
    / __ \/ / / / __ \   / ___/_________  ____  ___  _____
   / /_/ / /_/ / /_/ /   \__ \/ ___/ __ \/ __ \/ _ \/ ___/
  / ____/ __  / ____/   ___/ / /__/ /_/ / /_/ /  __/ /
 /_/   /_/ /_/_/       /____/\___/\____/ .___/\___/_/
                                      /_/

 PHP Scoper version 0.9.2

    0/631 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
   63/631 [▓▓░░░░░░░░░░░░░░░░░░░░░░░░░░]   9%
  126/631 [▓▓▓▓▓░░░░░░░░░░░░░░░░░░░░░░░]  19%
  189/631 [▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░░░░░░]  29%
  252/631 [▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░░░]  39%
  315/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░]  49%
  378/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░]  59%
  441/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░]  69%
  504/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░]  79%
  567/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░]  89%
  630/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░]  99%


  [OK] Successfully prefixed 631 files.

  // Memory usage: 17.79MB (peak: 34.99MB), time: 20.04s

  631/631 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
 ```
</details>

## Orignal
```php
<?php
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject
{
...
```

## php-scoper ouput
```php
<?php

namespace PHPUnit;

/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;
/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject
{
```
## Expected output
The expected output would be the original without any changes.

**Comment by theofidry** [comment]
Thanks for reporting the issue @sebastianfeldmann. Looking at it I don't really see any issue but it looks weird that in one case you have the namespace `PHPUnit` and not in the other.

Could you try to reproduce the issue with a breaking test? The format should be rather straightforward, you can see [this example](https://github.com/humbug/php-scoper/blob/master/specs/class-FQ.php#L28). Also to ease the testing, you can move the spec file under `_specs` instead of `specs`: `bin/phpunit tests/Scoper/PhpScoperSpecTest.php --filter test_can_scope_valid_files ` would then run only this spec file.

Also you might try `--stop-on-failure` when scoping the files to check that this file is not somehow skipped to due a failure (you can also see all failures by increasing the verbosity when prefixing the code)

**Comment by kambo-1st** [comment]
Side note - file still have a proper alias. This is on the end of the file:
```php
\class_alias('PHPUnit\\PHPUnit_Framework_MockObject_MockObject', 'PHPUnit_Framework_MockObject_MockObject', \false);
```
When i first saw this, i think, that this is the feature, not the issue.

**Comment by sebastianfeldmann** [comment]
This is getting spooky. After rebasing on the latest master of phpunit and re-running the phar generation, the unwanted namespace is still there but, the alias isn't added anymore.

I pushed my `php-scoper` integration on top of the latest master branch to https://github.com/sebastianfeldmann/phpunit/tree/feature/php-scoper

I removed the directory deletion in `build.xml` so after running `ant phar` you can examine the results in `build/phar` and `build/phar-scoped`.
So just the same as in @kambo-1st feature branch.

Steps to reproduce:
    $ cd /tmp
    $ git clone https://github.com/sebastianfeldmann/phpunit.git
    $ cd phpunit
    $ git fetch origin
    $ git checkout feature/php-scoper
    $ composer install
    $ ant phar

I'm still trying to create a smaller example or even a unit test to reproduce this.

**Comment by sebastianfeldmann** [comment]
I managed to reproduce this issue in a unit test.

specs/whitelist/whitelisted-global-namespace.php
```php
<?php

declare(strict_types=1);

/*
 * This file is part of the humbug/php-scoper package.
 *
 * Copyright (c) 2017 Théo FIDRY <theo.fidry@gmail.com>,
 *                    Pádraic Brady <padraic.brady@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

return [
    'meta' => [
        'title' => 'Whitelisted namespace',
        // Default values. If not specified will be the one used
        'prefix' => 'PHPUnit',
        'whitelist' => [],
        'whitelist-global-constants' => true,
        'whitelist-global-classes' => false,
        'whitelist-global-functions' => true,
        'registered-classes' => [],
        'registered-functions' => [],
    ],

    'Different whitelisted namespace types' => [
        'whitelist' => ['PHPUnit\\*', 'PHPUnit_Framework_MockObject_MockObject'],
        'registered-classes' => [],
        'payload' => <<<'PHP'
<?php
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject /*extends Verifiable*/
{
    /**
     * @return InvocationMocker
     */
    public function __phpunit_setOriginalObject($originalObject);

    /**
     * @return InvocationMocker
     */
    public function __phpunit_getInvocationMocker();

    /**
     * Verifies that the current expectation is valid. If everything is OK the
     * code should just return, if not it must throw an exception.
     *
     * @throws ExpectationFailedException
     */
    public function __phpunit_verify();

    /**
     * @return bool
     */
    public function __phpunit_hasMatchers();

    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);

    /**
     * Registers a new expectation in the mock object and returns the match
     * object which can be infused with further details.
     *
     * @return InvocationMocker
     */
    public function expects(Invocation $matcher);
}

----
<?php
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject /*extends Verifiable*/
{
    /**
     * @return InvocationMocker
     */
    public function __phpunit_setOriginalObject($originalObject);

    /**
     * @return InvocationMocker
     */
    public function __phpunit_getInvocationMocker();

    /**
     * Verifies that the current expectation is valid. If everything is OK the
     * code should just return, if not it must throw an exception.
     *
     * @throws ExpectationFailedException
     */
    public function __phpunit_verify();

    /**
     * @return bool
     */
    public function __phpunit_hasMatchers();

    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);

    /**
     * Registers a new expectation in the mock object and returns the match
     * object which can be infused with further details.
     *
     * @return InvocationMocker
     */
    public function expects(Invocation $matcher);
}

PHP
    ],
];

```

results in 

```
1) Humbug\PhpScoper\Scoper\PhpScoperSpecTest::test_can_scope_valid_files with data set #183 ('specs/different-namespace/glo...ce.php', '[Whitelisted namespace] Diffe... scope', '<?php\n/*\n * This file is pa...;\n}\n', 'PHPUnit', Humbug\PhpScoper\Whitelist Object (...), '<?php\n/*\n * This file is pa...;\n}\n', array(), array())
================================================================================
SPECIFICATION
================================================================================
[Whitelisted namespace] Different kind of whitelisted class constant calls in the global scope
specs/different-namespace/global-in-namespace.php

================================================================================
INPUT
whitelist: [
  - PHPUnit\*
  - PHPUnit_Framework_MockObject_MockObject
]
whitelist global constants: true
whitelist global functions: true
================================================================================
<?php
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject /*extends Verifiable*/
{
    /**
     * @return InvocationMocker
     */
    public function __phpunit_setOriginalObject($originalObject);

    /**
     * @return InvocationMocker
     */
    public function __phpunit_getInvocationMocker();

    /**
     * Verifies that the current expectation is valid. If everything is OK the
     * code should just return, if not it must throw an exception.
     *
     * @throws ExpectationFailedException
     */
    public function __phpunit_verify();

    /**
     * @return bool
     */
    public function __phpunit_hasMatchers();

    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);

    /**
     * Registers a new expectation in the mock object and returns the match
     * object which can be infused with further details.
     *
     * @return InvocationMocker
     */
    public function expects(Invocation $matcher);
}


================================================================================
EXPECTED
================================================================================
<?php
/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject /*extends Verifiable*/
{
    /**
     * @return InvocationMocker
     */
    public function __phpunit_setOriginalObject($originalObject);

    /**
     * @return InvocationMocker
     */
    public function __phpunit_getInvocationMocker();

    /**
     * Verifies that the current expectation is valid. If everything is OK the
     * code should just return, if not it must throw an exception.
     *
     * @throws ExpectationFailedException
     */
    public function __phpunit_verify();

    /**
     * @return bool
     */
    public function __phpunit_hasMatchers();

    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);

    /**
     * Registers a new expectation in the mock object and returns the match
     * object which can be infused with further details.
     *
     * @return InvocationMocker
     */
    public function expects(Invocation $matcher);
}

----------------
registered classes: []
registered functions: []

================================================================================
ACTUAL
================================================================================
<?php

namespace PHPUnit;

/*
 * This file is part of PHPUnit.
 *
 * (c) Sebastian Bergmann <sebastian@phpunit.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;
/**
 * Interface for all mock objects which are generated by
 * MockBuilder.
 *
 * @method InvocationMocker method($constraint)
 */
interface PHPUnit_Framework_MockObject_MockObject
{
    /**
     * @return InvocationMocker
     */
    public function __phpunit_setOriginalObject($originalObject);
    /**
     * @return InvocationMocker
     */
    public function __phpunit_getInvocationMocker();
    /**
     * Verifies that the current expectation is valid. If everything is OK the
     * code should just return, if not it must throw an exception.
     *
     * @throws ExpectationFailedException
     */
    public function __phpunit_verify();
    /**
     * @return bool
     */
    public function __phpunit_hasMatchers();
    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);
    /**
     * Registers a new expectation in the mock object and returns the match
     * object which can be infused with further details.
     *
     * @return InvocationMocker
     */
    public function expects(\PHPUnit\Framework\MockObject\Matcher\Invocation $matcher);
}

----------------
registered classes: [PHPUnit_Framework_MockObject_MockObject => PHPUnit\PHPUnit_Framework_MockObject_MockObject]
registered functions: []

-------------------------------------------------------------------------------
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 '<?php
+
+namespace PHPUnit;
+
 /*
  * This file is part of PHPUnit.
  *
@@ @@
  * For the full copyright and license information, please view the LICENSE
  * file that was distributed with this source code.
  */
-
 use PHPUnit\Framework\ExpectationFailedException;
 use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
 use PHPUnit\Framework\MockObject\Matcher\Invocation;
-
 /**
  * Interface for all mock objects which are generated by
  * MockBuilder.
@@ @@
  *
  * @method InvocationMocker method($constraint)
  */
-interface PHPUnit_Framework_MockObject_MockObject /*extends Verifiable*/
+interface PHPUnit_Framework_MockObject_MockObject
 {
     /**
      * @return InvocationMocker
      */
     public function __phpunit_setOriginalObject($originalObject);
-
     /**
      * @return InvocationMocker
      */
     public function __phpunit_getInvocationMocker();
-
     /**
      * Verifies that the current expectation is valid. If everything is OK the
      * code should just return, if not it must throw an exception.
@@ @@
      * @throws ExpectationFailedException
      */
     public function __phpunit_verify();
-
     /**
      * @return bool
      */
     public function __phpunit_hasMatchers();
-
     public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);
-
     /**
      * Registers a new expectation in the mock object and returns the match
      * object which can be infused with further details.
@@ @@
      *
      * @return InvocationMocker
      */
-    public function expects(Invocation $matcher);
+    public function expects(\PHPUnit\Framework\MockObject\Matcher\Invocation $matcher);
 }
 '

/Users/sf/Development/php/github.com/forks/php-scoper/tests/Scoper/PhpScoperSpecTest.php:130
```

**Comment by theofidry** [comment]
Sorry I cannot reproduce:

<details>
<summary>_specs/phpunit.php</summary>

```php
<?php declare(strict_types=1);

return [
    'meta' => [
        'title' => 'Whitelisted namespace',
        'prefix' => 'PHPUnit',
        'whitelist' => [],
        'whitelist-global-constants' => true,
        'whitelist-global-classes' => false,
        'whitelist-global-functions' => true,
        'registered-classes' => [],
        'registered-functions' => [],
    ],

    'Different whitelisted namespace types' => [
        'whitelist' => [
            'PHPUnit\\*',
            'PHPUnit_Framework_MockObject_MockObject'
        ],
        'registered-classes' => [
            [
                'PHPUnit_Framework_MockObject_MockObject',
                'PHPUnit\PHPUnit_Framework_MockObject_MockObject'
            ],
        ],
        'payload' => <<<'PHP'
<?php

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;

interface PHPUnit_Framework_MockObject_MockObject
{
    public function __phpunit_setOriginalObject($originalObject);
    public function __phpunit_getInvocationMocker();
    public function __phpunit_verify();
    public function __phpunit_hasMatchers();
    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);
    public function expects(Invocation $matcher);
}

----
<?php

namespace PHPUnit;

use PHPUnit\Framework\ExpectationFailedException;
use PHPUnit\Framework\MockObject\Builder\InvocationMocker;
use PHPUnit\Framework\MockObject\Matcher\Invocation;
interface PHPUnit_Framework_MockObject_MockObject
{
    public function __phpunit_setOriginalObject($originalObject);
    public function __phpunit_getInvocationMocker();
    public function __phpunit_verify();
    public function __phpunit_hasMatchers();
    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration);
    public function expects(\PHPUnit\Framework\MockObject\Matcher\Invocation $matcher);
}

PHP
    ],
];
```
</details>

As you can see you have the whitelisted class properly recorded:

```
'registered-classes' => [
            [
                'PHPUnit_Framework_MockObject_MockObject',
                'PHPUnit\PHPUnit_Framework_MockObject_MockObject'
            ],
        ],
```

Which guarantees the `PHPUnit_Framework_MockObject_MockObject` will have the alias `PHPUnit\PHPUnit_Framework_MockObject_MockObject` with the `class_alias()` statement in the `scoper-autoload.php` file

**Comment by sebastianfeldmann** [comment]
What I wanted to achieve was, that `PHPUnit_Framework_MockObject_MockObject` will not be changed at all. No additional namespace, no alias.
That's what I expected whitelisting the class would result in.

**Comment by sebastianfeldmann** [comment]
OK I understand (... I think ...)
I will try to figure out a way to execute the class alias in my scenario.

Thanks for your help

**Comment by theofidry** [comment]
>I will try to figure out a way to execute the class alias in my scenario.

Under what circumstances the alias is not triggered?

Be careful that you need to autoload `vendor/scoper-autoload.php` instead of `vendor/autoload.php` if you are using whitelisting features. I handle that transparently in [Box](https://github.com/humbug/box) but not in PHP-Scoper (that could be done but I don't have the time to port it)

**Comment by sebastianfeldmann** [comment]
In case of the `PHPUnit` PHAR the composer autoloader isn't used. `PHPUnit` uses `phpab` to create the PHAR file which uses a different mechanism to load all classes inside the PHAR.

Essentially how it works is:
You provide an directory and `phpab` includes each file in the PHAR and adds a require statement for it to its "autoloader".

**Comment by theofidry** [comment]
Although the wall discussion still stands, it looks like there was indeed a bug which has been fixed in #263
